[memprof] Reduce schema for Version2

kazutakahirata opened the pull request:
Curently, the compiler only uses several fields of MemoryInfoBlock.
Serializing all fields into the indexed MemProf file simply wastes
storage.

This patch limits the schema down to four fields for Version2 by
default.  It retains the old behavior of serializing all fields via:

  llvm-profdata merge --memprof-version=2 --memprof-full-schema

This patch reduces the size of the indexed MemProf profile I have by
40% (1.6GB down to 1.0GB).

kazutakahirata committed:
[memprof] Reduce schema for Version2

Curently, the compiler only uses several fields of MemoryInfoBlock.
Serializing all fields into the indexed MemProf file simply wastes
storage.

This patch limits the schema down to four fields for Version2 by
default.  It retains the old behavior of serializing all fields via:

  llvm-profdata merge --memprof-version=2 --memprof-full-schema

This patch reduces the size of the indexed MemProf profile I have by
40% (1.6GB down to 1.0GB).

teresajohnson reviewed (commented):
Question about testing below.

It also occurs to me that we should do something on the profile use side in the compiler to detect which fields were in the profile, and give an error if we try to use other fields. I have a bit of concern that if someone tries to experiment with compiler changes on the use side that make use of additional fields and forgets about this change that they will get 0s for other fields and it silently won't behave as expected. E.g. maybe in llvm/include/llvm/ProfileData/MemProf.h possibly save the Schema used in deserialize, and assert in the get* functions if a field is used that wasn't in that set?

snehasish commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -117,14 +117,21 @@ struct PortableMemInfoBlock {
 
+  // Returns the schema consisting of the fields currently consumed by the
+  // compiler.
+  static MemProfSchema getSchema() {
```
Maybe we should qualify the schema based on the use case. For example these fields are necessary for hot-cold annotation, so name this `getHotColdSchema()`?

kazutakahirata replied:
I'll incorporate the suggested name in this patch.  Thanks!

snehasish commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -117,14 +117,21 @@ struct PortableMemInfoBlock {
 
   // Returns the full schema currently in use.
-  static MemProfSchema getSchema() {
+  static MemProfSchema getFullSchema() {
```
Any reason that these are part of the PortableMemInfoBlock? IMO they should be peers of the Schema itself (under llvm::memprof).

kazutakahirata replied:
I'll post a separate patch for that.

kazutakahirata commented on llvm/test/tools/llvm-profdata/memprof-merge-v0.test:
```diff
@@ -16,6 +16,9 @@ RUN: llvm-profdata show %t.prof.v1 | FileCheck %s
 RUN: llvm-profdata show %t.prof.v2 | FileCheck %s
 
+RUN: llvm-profdata merge %t.proftext %p/Inputs/basic.memprofraw --memprof-version=2 --memprof-full-schema --profiled-binary %p/Inputs/basic.memprofexe -o %t.prof.v2
+RUN: llvm-profdata show %t.prof.v2 | FileCheck %s
```
> Is there any way to check the output to confirm that we either serialized 4 fields or all fields?

Yes, let me try to extend `llvm/unittests/ProfileData/InstrProfTest.cpp` to verify that the roundtrip works in each case -- the 4 fields or all fields.

> Also, just noticed that this test has "v0" in the name, should the name be changed to reflect the fact that it is testing all versions? (not in this PR but separately)

Yes, I just borrowed a test `memprof-merge.test` and added `-v0` to it.  I'll come up with a better name in a separate patch.

snehasish commented on llvm/lib/ProfileData/InstrProfWriter.cpp:
```diff
@@ -563,7 +565,11 @@ static Error writeMemProfV2(
   OS.write(0ULL); // Reserve space for the memprof call stack table offset.
 
-  auto Schema = memprof::PortableMemInfoBlock::getSchema();
+  memprof::MemProfSchema Schema;
```
nit: Just initialize with the one we want and then override.
```
memprof::MemProfSchema Schema = memprof::PortableMemInfoBlock::getHotColdSchema();
if (MemProfFullSchema)
  Schema = memprof::PortableMemInfoBlock::getFullSchema();
```

snehasish commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -308,6 +308,11 @@ cl::opt<memprof::IndexedVersion> MemProfVersionRequested(
 
+cl::opt<bool> MemProfFullSchema("memprof-full-schema", cl::Hidden,
+                                cl::sub(MergeSubcommand),
+                                cl::desc("Serialize the full schema"),
```
nit: "Use the full schema for serialization" -- The current statement sounds like it applies to the schema and not the content.

kazutakahirata replied:
Adressed in the latest iteration.

kazutakahirata commented:
> It also occurs to me that we should do something on the profile use side in the compiler to detect which fields were in the profile, and give an error if we try to use other fields. I have a bit of concern that if someone tries to experiment with compiler changes on the use side that make use of additional fields and forgets about this change that they will get 0s for other fields and it silently won't behave as expected. E.g. maybe in llvm/include/llvm/ProfileData/MemProf.h possibly save the Schema used in deserialize, and assert in the get* functions if a field is used that wasn't in that set?

I understand your concern.  Yes, `get*` functions seem to be the way to go.  Let me think about the details.

snehasish commented:
lgtm, pending the unittest for the schema based roundtrip.

kazutakahirata committed:
Address comments.

- Rename getSchema to getHotColdSchema.

- Add a unit test for the full/partial schema.

- Update the description for the memprof-full-schema option.

kazutakahirata committed:
Update the call to getHotColdSchema.

snehasish commented on llvm/unittests/ProfileData/InstrProfTest.cpp:
```diff
@@ -370,12 +370,53 @@ static CallStackIdMapTy getCallStackMapping() {
 }
 
+// Populate all of the fields of MIB.
+MemInfoBlock makeFullMIB() {
```
I think we should use the macro expansion of fields here so that it's future proof i.e. any new fields added to the original MIBEntry will automatically get picked up.

snehasish reviewed (approved):
New unit test lgtm with one minor comment. Thanks!

kazutakahirata committed:
Teach makeFullMIB to use MIBEntryDef.inc.
